cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java

```diff
@@ -586,19 +594,32 @@ private ModelCompletenessRequirements modelCompletenessRequirements(Collection<G
    * @param goalNames Goal names (and empty list of names indicates all goals).
    */
   public boolean meetCompletenessRequirements(List<String> goalNames) {
-    Collection<Goal> goals = goalsByPriority(goalNames).values();
+    Collection<Goal> goals = goalsByPriority(goalNames, false).values();
     MetadataClient.ClusterAndGeneration clusterAndGeneration = _loadMonitor.refreshClusterAndGeneration();
     return goals.stream().allMatch(g -> _loadMonitor.meetCompletenessRequirements(
         clusterAndGeneration, g.clusterModelCompletenessRequirements()));
   }
 
   /**
    * Get a goals by priority based on the goal list.
+   *
+   * @param goals A list of goals to optimize.
+   * @param skipHardGoalCheck Whether check and add all hard goals to {@code goals}.
+   * @return A map of goal priority to goal.
    */
-  private Map<Integer, Goal> goalsByPriority(List<String> goals) {
+  private Map<Integer, Goal> goalsByPriority(List<String> goals, boolean skipHardGoalCheck) {
     if (goals == null || goals.isEmpty()) {
       return AnalyzerUtils.getGoalMapByPriority(_config);
     }
+    if (!skipHardGoalCheck) {
+      List<String> hardGoals = _config.getList(KafkaCruiseControlConfig.HARD_GOALS_CONFIG);
+      for (String goalName : goals) {
+        if (!hardGoals.contains(goalName)) {
+          hardGoals.add(goalName);
+        }
+      }
+      goals = hardGoals;
```

Comment:
I think we should avoid silently overwriting the user parameters. The role of the `skipHardGoalCheck ` should be: `If the user does not explicitly ask to skip the hard goal check, then if CC identifies that a hard goal is missing from the user provided goals, then throw an exception.`